New rule ensure_chrony_is_configured

rumch-se opened the pull request:
#### Description:

- _Rule covers CIS requirement 2.2.1.3 Ensure chrony is configured (Automated)_

#### Rationale:

- The rule has audit and remediation part which check/add remote NTP servers to /etc/chrony.conf

dodys commented on linux_os/guide/services/ntp/ensure_chrony_is_configured/oval/shared.xml:
```diff
@@ -0,0 +1,20 @@
+  <ind:textfilecontent54_object comment="Ensure at least one remote NTP server is set"
+  id="object_chrony_remote_server_configuartion" version="1">
+    <ind:filepath operation="pattern match">^/etc/chrony.conf$</ind:filepath>
+    <ind:pattern operation="pattern match">^(\bserver[[:space:]]|pool[[:space:]]\b)([0-3]\b(.fedora|.suse|.rhel)\b\b(.pool.ntp.org)\b)|([0-3]\b(.pool.ntp.org)\b)|([0-3]\b(.ntp.cloud.aliyuncs.com)\b)|([0-3]\b(.us.pool.ntp.mil)\b)$</ind:pattern>
```
shouldn't you be checking if the server specified in `var_multiple_time_servers`? 
That would allow anyone working with a tailoring file to set `var_multiple_time_servers` to the servers according to their site policy, and that might be different from this hard-coded check.

dodys replied:
Also isn't this entire rule the same as `chronyd_specify_remote_server`?

dodys replied:
I wouldn't use any regex, I would check if the `/etc/chrony.conf` file has the value set in `var_multiple_time_servers`, that way you allow users to define their own servers.

rumch-se replied:
Also @dodys 
According to your second question I will propose you to check conversation in https://github.com/ComplianceAsCode/content/pull/10278. According to me the OVAL part of the rule chronyd_specify_remote_server is not correct - because it checks for ANY word after the word server. I received a proposal to develop another rule, and because of that this PR was developed. With this PR I am trying to control better the content of /etc/chrony.conf - because the correct configuration of servers is very important for audit/ digital forensic/ how IDS/IPS are working.  
Have a nice day
Rumen

dodys commented on linux_os/guide/services/ntp/ensure_chrony_is_configured/oval/shared.xml:
```diff
@@ -0,0 +1,20 @@
+
+  <ind:textfilecontent54_object comment="Ensure at least one remote NTP server is set"
+  id="object_chrony_remote_server_configuartion" version="1">
+    <ind:filepath operation="pattern match">^/etc/chrony.conf$</ind:filepath>
```
I believe here you might want to use: {{{ chrony_conf_path }}}
see other rules that have this variable use

dodys commented:
> Hello @dodys I did necessary changes Have a nice day Rumen

hi @rumch-se
you've added four new variables when you already have the var_multiple_time_servers.  I don't think you need that at all, you can have a local_variable in oval that is generated from the var_multiple_time_servers. No need for new variables outside of oval.
I'm doing some tests and will propose a better solution

dodys commented:
Hi @rumch-se, I think I got a working prototype of the OVAL.

Do note that I did on top of `chronyd_specify_remote_server` just because that's what we have in master right now, but you could re-use it in your new rule.

1. Here is the diff I added to the oval in `chronyd_specify_remote_server`: https://pastebin.ubuntu.com/p/CV4PC9YTqc/
    1.1. To explain what it is doing, first I'm referencing the external_variable `var_multiple_time_servers`, which in my case here is set to the default value of `0.pool.ntp.org,1.pool.ntp.org,2.pool.ntp.org,3.pool.ntp.org`
    1.2. I'm creating a local_variable that is first splitting the value of `var_multiple_time_servers` by the ','. From this split you will get 4 values: `0.pool.ntp.org`, `1.pool.ntp.org`, `2.pool.ntp.org` and `3.pool.ntp.org`. Then it concatenates the string `^(?:server|pool)[[:space:]]` with each of the values from the split + the string `$`. That might be confusing but what happens in oval is de following:
``` 
I: oscap:     Variable 'oval:ssg-temp_variable_test:var:1' has values
 "^(?:server|pool)[[:space:]]0.pool.ntp.org$",
 "^(?:server|pool)[[:space: ]]1.pool.ntp.org$",
 "^(?:server|pool)[[:space:]]2.pool.ntp.org$",
 "^(?:server|pool)[[:space:]]3.pool.ntp.org$".
 [oscap(1934): oscap(7fba75ff0980):oval_variable.c:513:_dump_variable_values]
```
  
 2. Now that we have the local_variable setup, I just changed `textfilecontent54_object` to actually:
     2.1. use `{{{ chrony_conf_path }}}`  just to keep consistency with the other rules
     2.2. make the pattern reference the local_variable we created, after all it is a pattern
     2.3. altered the instance check for more than one instance is a match, since we expect 4 matches.
     
 This is the result I got:
 
![image](https://user-images.githubusercontent.com/1580168/227561807-13261305-5377-44b3-b1ea-b2f316212771.png)

You can see it correctly matched the four entries in chrony.conf

dodys commented:
@vojtapolasek I wonder if we can have the OVAL change I mentioned above in `chronyd_specify_remote_server` and not need this new rule at all. Thoughts?

rumch-se commented:
Hello @dodys 

I have corrected this rule, by following your approach. In addition I have added support for server and pool, because: 
- the server directive specifies an NTP server which can be used as a time source
- the pool directive specifies a pool of NTP servers rather a single NTP server. The pool name is expected to resolve to multiple addresses which might change over time.
I have defined a variable for pool.
 
Have a nice day
Rumen

dodys commented on linux_os/guide/services/ntp/ensure_chrony_is_configured/oval/shared.xml:
```diff
@@ -0,0 +1,58 @@
+
+  <ind:textfilecontent54_test check="all" check_existence="at_least_one_exists"
+  comment="Ensure remote NTP server is set " id="test_chrony_remote_server_configuration" version="1">
+    <ind:object object_ref="object_chrony_remote_server_configuartion" />
```
typo in configuration

dodys commented on linux_os/guide/services/ntp/ensure_chrony_is_configured/oval/shared.xml:
```diff
@@ -0,0 +1,58 @@
+    <ind:object object_ref="object_chrony_remote_server_configuartion" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object comment="Ensure at least one NTP server is set" id="object_chrony_remote_server_configuartion" version="1">
```
typo in configuration

dodys commented on linux_os/guide/services/ntp/ensure_chrony_is_configured/oval/shared.xml:
```diff
@@ -0,0 +1,58 @@
+  </local_variable>
+
+  <ind:textfilecontent54_test check="all" check_existence="at_least_one_exists"
+  comment="Ensure remote NTP server is set " id="test_chrony_remote_server_configuration" version="1">
```
please remove the space after "set" in comment

dodys commented on linux_os/guide/services/ntp/ensure_chrony_is_configured/oval/shared.xml:
```diff
@@ -0,0 +1,58 @@
+
+  <ind:textfilecontent54_test check="all" check_existence="at_least_one_exists"
+  comment="Ensure remote NTP pool is set " id="test_chrony_remote_pool_configuration" version="1">
+    <ind:object object_ref="object_chrony_remote_pool_configuartion" />
```
typo in configuration

dodys reviewed (changes requested):
a few minor things to fix, but overall it is looking good now

rumch-se commented:
Hello @dodys 
The necessary corrections were done.
Have a nice day
Rumen

marcusburghardt commented:
Related to https://github.com/ComplianceAsCode/content/pull/10278

dodys commented:
> Related to #10278

@marcusburghardt @vojtapolasek is there any reason for this new rule to not be integrated with the existing rule?

mildas commented:
/packit test

vojtapolasek commented:
@rumch-se  I am very sorry for a long delay, I am going to review this PR asap since I know it is important for you.
@dodys  the rule was not integrated into chronyd_specify_remote_server because I believe it introduces a big change to the rule from user's point of view.
I consider the case when user decices to use some time servers which are not configured in the XCCDF variable, for example their internal corporate servers. In that case the rule would start failing out of nowhere (if it was passing before). To make rule passing user needs to modify the XCCDF variable. I do not dispute that the rule in this PR is more secure and better, but I don't think we should simply replace the rule chronyd_specify_remote_server by this rule because of user experience. However, we might gradually transition to the new rule.

vojtapolasek reviewed (changes requested):
Hello @rumch-se  and thank you for this new rule. It looks challenging but I believe the goal you are trying to accomplish can be done.
See my comments please. Please fix the wrong variable name in the pci-dss profile so that the product can be built and tested by the CI.
I have two general comments:
1. Could you consider a different rule name? This name sounds very generic - I believe time servers and pools are not the only configuration options which can be set in chrony.conf. How about chrony_configure_exact_servers... or something along the lines.
2. I see that the rule remediation (I tried Bash so far) configures both servers and pools in the file. Is this what you want to achieve?
Thank you once again.

rumch-se commented:
Hello @vojtapolasek 

I have implemented the proposed corrections.

Have a nice day
Rumen

vojtapolasek commented on linux_os/guide/services/ntp/chronyd_configures_time_sources/ansible/shared.yml:
```diff
@@ -0,0 +1,25 @@
+- name: {{{ rule_title }}} - Add missing / update wrong records for remote time pools
+  ansible.builtin.lineinfile:
+    path: {{{ chrony_conf_path }}}
+    regexp: '^\s*\pool\b\s*\b{{ item }}\b$'
```
```suggestion
    regexp: '^\s*\bpool\b\s*\b{{ item }}\b$'
```

vojtapolasek commented on linux_os/guide/services/ntp/chronyd_configures_time_sources/ansible/shared.yml:
```diff
@@ -0,0 +1,25 @@
+    path: {{{ chrony_conf_path }}}
+    regexp: '^\s*\pool\b\s*\b{{ item }}\b$'
+    state: present
+    line: 'pool {{ item }}'
```
Please add create: true because if the file does not exist the task fails.

vojtapolasek commented on linux_os/guide/services/ntp/chronyd_configures_time_sources/ansible/shared.yml:
```diff
@@ -0,0 +1,25 @@
+    path: {{{ chrony_conf_path }}}
+    regexp: '^\s*\bserver\b\s*\b{{ item }}\b$'
+    state: present
+    line: 'server {{ item }}'
```
Please add create: true because if the file does not exist the task fails.

vojtapolasek reviewed (changes requested):
Thank you for the update. There are just minor fixes to Ansible remaining because in the current state it ofter terminates prenaturely.

rumch-se commented:
Hello @vojtapolasek 

The proposed changes were implemented 
Have a nice day
Rumen

vojtapolasek reviewed (approved):
Thank you for the update. It looks good now.
Automatus tests are failing because the rule is SLE specific and therefore Automats has nothing to test... and returns error.

vojtapolasek commented:
@teacup-on-rockingchair  could you please review?

vojtapolasek commented:
@dodys  and @ComplianceAsCode/suse-maintainers  please review the PR. Thank you.

vojtapolasek commented:
@rumch-se  one more thing to consider... what do you think about splitting the rule into two rules? One would check for pools, one for servers. The reason why I am asking is for example this STIG item:
https://stigaview.com/products/rhel8/v1r9/RHEL-08-030740/
It does not mention pools at all, only the server directive.

rumch-se commented:
Hello @vojtapolasek
What I know about server and pool is the following. In case of server you will have 1 source of time, in case of pool - you will have a set of servers. In other words, the pool is more robust - because if one of time servers is not active, the remaining ones will provide time. When I created the rule I used CIS recommendation where we have server and pool.
Have a nice day
Rumen

vojtapolasek commented:
I understand. I think it would be nice to name the rule chronyd_configure_server_and_pool. We could then have separate rules (out of scope of this PR) chronyd_configure_server and chronyd_configure_pool.  @yuumasato  @jan-cerny  ideas?

teacup-on-rockingchair reviewed (approved):
LGTM :+1:

rumch-se commented:
Hello @vojtapolasek 
I changed the name of the rule.
Have a nice day
Rumen

vojtapolasek commented:
Thank you @rumch-se, merging.
